File: src/main/java/org/example/bankramenserver/domain/auth/client/KakaoOAuthClient.java
package org.example.bankramenserver.domain.auth.client;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.example.bankramenserver.domain.auth.dto.response.KakaoUserResponse;
import org.example.bankramenserver.domain.auth.exception.KaKaoUserInfoRequestFailedException;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.*;
import org.springframework.stereotype.Component;
import org.springframework.web.client.RestTemplate;

@Slf4j
@Component
@RequiredArgsConstructor
public class KakaoOAuthClient {

    private final RestTemplate restTemplate;

    @Value("${kakao.api-url}")
    private String apiUrl;

    public KakaoUserResponse requestUserInfo(String kakaoAccessToken) {
        HttpHeaders headers = new HttpHeaders();
        headers.setBearerAuth(kakaoAccessToken);

        try {
            KakaoUserResponse response = restTemplate.exchange(
                    apiUrl + "/v2/user/me",
                    HttpMethod.GET,
                    new HttpEntity<>(headers),
                    KakaoUserResponse.class
            ).getBody();

            if (response == null || response.id() == null) {
                throw KaKaoUserInfoRequestFailedException.EXCEPTION;
            }

            return response;
        } catch (Exception e) {
            log.error("카카오 사용자 정보 요청 실패", e);
            throw KaKaoUserInfoRequestFailedException.EXCEPTION;
        }
    }
}

File: src/main/kotlin/org/example/bankramenserver/domain/auth/client/KakaoOAuthClient.kt
package org.example.bankramenserver.domain.auth.client
import org.example.bankramenserver.domain.auth.dto.response.KakaoUserResponse
import org.example.bankramenserver.domain.auth.exception.KaKaoUserInfoRequestFailedException
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Value
import org.springframework.http.HttpEntity
import org.springframework.http.HttpHeaders
import org.springframework.http.HttpMethod
import org.springframework.stereotype.Component
import org.springframework.web.client.RestTemplate
@Component class KakaoOAuthClient(private val restTemplate: RestTemplate, @Value("\${kakao.api-url}") private val apiUrl: String) { private val log = LoggerFactory.getLogger(javaClass); fun requestUserInfo(kakaoAccessToken: String): KakaoUserResponse = try { val headers = HttpHeaders().also { it.setBearerAuth(kakaoAccessToken) }; restTemplate.exchange("$apiUrl/v2/user/me", HttpMethod.GET, HttpEntity<Void>(headers), KakaoUserResponse::class.java).body?.takeIf { it.id != null } ?: throw KaKaoUserInfoRequestFailedException.EXCEPTION } catch (e: Exception) { log.error("카카오 사용자 정보 요청 실패", e); throw KaKaoUserInfoRequestFailedException.EXCEPTION } }
